validator/sawtooth_validator/journal/publisher.py

```diff
@@ -32,10 +32,200 @@
+        else:
+            LOGGER.debug("Dropping batch due to missing dependencies: %s",
+                         batch.header_signature)
+
+    def check_publish_block(self):
+        return self._consensus.check_publish_block(
+            self._block_builder.block_header)
+
+    def _sign_block(self, block, identity_signing_key):
+        """ The block should be complete and the final
+        signature from the publishing validator(this validator) needs to
+        be added.
+        :param block: the Block to sign.
+        :param identity_signing_key: the key to sign the block with.
+        """
+        header_bytes = block.block_header.SerializeToString()
+        signature = signing.sign(header_bytes, identity_signing_key)
+        block.set_signature(signature)
+
+    def finalize_block(self, identity_signing_key, pending_batches):
+        """Compose the final Block to publish. This involves flushing
+        the scheduler, having consensus bless the block, and signing
+        the block.
+        :param identity_signing_key: the key to sign the block with.
+        :param pending_batches: list to receive any batches that were
+        submitted to add to the block but were not validated before this
+        call.
+        """
+        self._scheduler.finalize()
+        self._scheduler.complete(block=True)
+
+        # this is a transaction cache to track the transactions committed
+        # up to this batch. Only valid transactions that were processed
+        # by the scheduler are added.
+        committed_txn_cache = TransactionCache(self._block_store)
+
+        builder = self._block_builder
+        state_hash = None
+        for batch in self._pending_batches:
+            result = self._scheduler.get_batch_execution_result(
+                batch.header_signature)
+            # if a result is None, this means that the executor never
+            # received the batch and it should be added to
+            # the pending_batches, to be added to the next
+            # block
+            if result is None:
+                pending_batches.append(batch)
+            elif result.is_valid:
+                # check if a dependent batch failed. This could be belt and
+                # suspenders action here but it is logically possible that
+                # a transaction has a dependency that fails it could
+                # still succeed validation. In which case we do not want
+                # to add it to the batch.
+                if not self._check_batch_dependencies(batch,
+                                                      committed_txn_cache):
+                    LOGGER.debug("Batch %s invalid, due to missing txn "
+                                 "dependency.", batch.header_signature)
+                    LOGGER.debug("Abandoning block %s:" +
+                                 "root state hash has invalid txn applied",
+                                 builder)
+                    pending_batches.remove(batch)
```

Comment:
When running locally with `sawtooth.poet.target_wait_time=5` and `sawtooth.poet.initial_wait_time=1`, got the following:
```
validator-2_1              | [17:25:10.247 INFO     chain] Starting block validation of : b6114567(34, S:82d27cee, P:c1d71581)
validator-2_1              | [17:25:10.651 INFO     poet_fork_resolver] Choose new fork b6114567: New fork aggregate local mean (170.000000) greater than current fork aggregate local mean (165.000000)
validator-2_1              | [17:25:10.673 INFO     chain] on_block_validated: b6114567(34, S:82d27cee, P:c1d71581)
validator-2_1              | [17:25:10.676 INFO     chain] Chain head updated to: b6114567(34, S:82d27cee, P:c1d71581)
validator-2_1              | [17:25:10.736 CRITICAL publisher] on_check_publish_block exception.
validator-2_1              | [17:25:10.737 ERROR    publisher] list.remove(x): x not in list
validator-2_1              | Traceback (most recent call last):
validator-2_1              |   File "/project/sawtooth-core/validator/sawtooth_validator/journal/publisher.py", line 453, in on_check_publish_block
validator-2_1              |     pending_batches)
validator-2_1              |   File "/project/sawtooth-core/validator/sawtooth_validator/journal/publisher.py", line 227, in finalize_block
validator-2_1              |     pending_batches.remove(batch)
validator-2_1              | ValueError: list.remove(x): x not in list
```
I can provide full log (although it is not a debug log) if you want.